File: packages/core/src/types/parsePlan.ts
import { z } from "zod";

const NAME_RE = /^[a-z][a-z0-9_]*$/;

/**
 * Compile a plan-authored row-rule / unpivot pattern. Matching is always
 * case-insensitive (the "i" flag), so a leading PCRE inline-flag group like
 * `(?i)` — which LLM proposers commonly emit but ECMAScript regex rejects as a
 * syntax error — is redundant and stripped before compilation. Throws (like
 * `new RegExp`) on a genuinely malformed pattern so validateParsePlan still
 * reports it.
 */
export function planPattern(pattern: string): RegExp {
  return new RegExp(pattern.replace(/^\(\?[a-z]+\)/i, ""), "i");
}

export const RowRuleSchema = z.object({
  /** Canonical column name, or null = match against any matched cell in the row. */
  column: z.string().nullable(),
  /** Case-insensitive regex tested against the cell's raw (trimmed) text. */
  pattern: z.string(),
});
export type RowRule = z.infer<typeof RowRuleSchema>;

export const ColumnPlanSchema = z.object({
  /** Normalized merged raw header text this column matches. */
  from: z.string().min(1),
  name: z.string().regex(NAME_RE),
  type: z.enum(["TEXT", "INTEGER", "REAL"]),
  parse: z.enum(["number", "currency", "percent", "date"]).optional(),
});
export type ColumnPlan = z.infer<typeof ColumnPlanSchema>;

export const TablePlanSchema = z.object({
  name: z.string().regex(NAME_RE),
  anchor: z.object({
    /** Slugified sheet-name HINT; breaks ties, never required to match. */
    sheet: z.string().optional(),
    /** Normalized raw texts of the header row's cells. */
    headerSignature: z.array(z.string().min(1)).min(1),
    minMatch: z.number().int().min(1),
  }),
  headerRows: z.number().int().min(1).max(3),
  exclude: z.array(RowRuleSchema),
  columns: z.array(ColumnPlanSchema).min(1),
  unpivot: z
    .object({
      keep: z.array(z.string()),
      /** Case-insensitive regex; matched (unmapped) merged headers melt into rows. */
      valuePattern: z.string(),
      keyColumn: z.string().regex(NAME_RE),
      valueColumn: z.string().regex(NAME_RE),
      valueType: z.enum(["INTEGER", "REAL", "TEXT"]),
      valueParse: z.enum(["number", "currency", "percent"]).optional(),
    })
    .optional(),
  /** Canonical column (parse: "date") validated against the slot period. */
  periodColumn: z.string().optional(),
  onPeriodMismatch: z.enum(["keep", "exclude"]).default("keep"),
});
export type TablePlan = z.infer<typeof TablePlanSchema>;

export const ParsePlanSchema = z.object({
  version: z.literal(1),
  tables: z.array(TablePlanSchema).min(1),
});
export type ParsePlan = z.infer<typeof ParsePlanSchema>;

export const ExecReportSchema = z.object({
  tables: z.array(
    z.object({
      name: z.string(),
      anchor: z.object({ sheet: z.string(), row: z.number() }).nullable(),
      /** Byte-exact problem lines; a table with problems produces no rows. */
      problems: z.array(z.string()),
      rowsKept: z.number(),
      rowsExcluded: z.array(z.object({ pattern: z.string(), count: z.number(), samples: z.array(z.string()) })),
      coercionFailures: z.array(z.object({ column: z.string(), count: z.number(), samples: z.array(z.string()) })),
      periodMismatches: z.object({ count: z.number(), samples: z.array(z.string()) }).nullable(),
      unknownColumns: z.array(z.string()),
    }),
  ),
});
export type ExecReport = z.infer<typeof ExecReportSchema>;

export const PlanPreviewSchema = z.object({
  tables: z.array(
    z.object({
      name: z.string(),
      columns: z.array(z.string()),
      rows: z.array(z.array(z.union([z.string(), z.number(), z.null()]))),
    }),
  ),
});
export type PlanPreview = z.infer<typeof PlanPreviewSchema>;

/**
 * Structural validity beyond zod. Throws Error whose message is the first
 * violated rule; messages are stable and user-visible in tool errors.
 */
export function validateParsePlan(plan: ParsePlan): void {
  const tableNames = new Set<string>();
  for (const t of plan.tables) {
    if (tableNames.has(t.name)) throw new Error(`duplicate table name: ${t.name}`);
    tableNames.add(t.name);
    const canon = new Set<string>();
    const froms = new Set<string>();
    for (const c of t.columns) {
      if (c.name === "_period") throw new Error(`reserved column name: ${t.name}._period`);
      if (canon.has(c.name)) throw new Error(`duplicate column name: ${t.name}.${c.name}`);
      canon.add(c.name);
      const f = c.from.trim().toLowerCase().replace(/\s+/g, " ");
      if (froms.has(f)) throw new Error(`duplicate column from: ${t.name}.${c.from}`);
      froms.add(f);
    }
    const ref = (col: string | null, where: string): void => {
      if (col !== null && !canon.has(col)) throw new Error(`unknown column reference: ${t.name}.${col}`);
      void where;
    };
    for (const r of t.exclude) {
      ref(r.column, "exclude");
      try { planPattern(r.pattern); } catch { throw new Error(`invalid pattern: ${t.name}.${r.pattern}`); }
    }
    if (t.unpivot) {
      for (const k of t.unpivot.keep) ref(k, "keep");
      try { planPattern(t.unpivot.valuePattern); } catch { throw new Error(`invalid pattern: ${t.name}.${t.unpivot.valuePattern}`); }
      if (canon.has(t.unpivot.keyColumn) || canon.has(t.unpivot.valueColumn) || t.unpivot.keyColumn === t.unpivot.valueColumn)
        throw new Error(`unpivot column collides: ${t.name}`);
    }
    if (t.periodColumn !== undefined) {
      ref(t.periodColumn, "periodColumn");
      const c = t.columns.find((x) => x.name === t.periodColumn);
      if (!c || c.parse !== "date") throw new Error(`periodColumn must have parse "date": ${t.name}.${t.periodColumn}`);
    }
  }
}

/** Warehouse table name for one logical table (mirrors tableNamesFor's single-table rule). */
export function planTableName(familyKey: string, plan: ParsePlan, logicalName: string): string {
  return plan.tables.length === 1 ? `fam_${familyKey}` : `fam_${familyKey}__${logicalName}`;
}

File: backend/runoff_api/core/types/parse_plan.py
"""Port of packages/core/src/types/parsePlan.ts.

zod objects here are plain (non-strict) -> CamelModelOpen. `validate_parse_plan`
and `plan_table_name` operate on plain dicts (the runtime shape), mirroring the
TS functions that receive already-parsed ParsePlan objects.
"""

import re
from typing import Annotated, Literal

from pydantic import Field

from .base import CamelModelOpen

NAME_RE = r"^[a-z][a-z0-9_]*$"


def plan_pattern(pattern: str) -> re.Pattern:
    """Port of planPattern.

    Matching is always case-insensitive, so a leading PCRE inline-flag group like
    `(?i)` (which LLM proposers emit but Python/ECMAScript reject) is stripped
    before compilation. Raises re.error (like `new RegExp`) on a malformed pattern.
    """
    return re.compile(re.sub(r"^\(\?[a-z]+\)", "", pattern, flags=re.IGNORECASE), re.IGNORECASE)


class RowRule(CamelModelOpen):
    # Canonical column name, or None = match against any matched cell in the row.
    column: str | None
    # Case-insensitive regex tested against the cell's raw (trimmed) text.
    pattern: str


class ColumnPlan(CamelModelOpen):
    # Normalized merged raw header text this column matches.
    from_: str = Field(alias="from", min_length=1)
    name: str = Field(pattern=NAME_RE)
    type: Literal["TEXT", "INTEGER", "REAL"]
    parse: Literal["number", "currency", "percent", "date"] | None = None


class TableAnchor(CamelModelOpen):
    # Slugified sheet-name HINT; breaks ties, never required to match.
    sheet: str | None = None
    # Normalized raw texts of the header row's cells.
    header_signature: list[Annotated[str, Field(min_length=1)]] = Field(min_length=1)
    min_match: int = Field(ge=1)


class Unpivot(CamelModelOpen):
    keep: list[str]
    # Case-insensitive regex; matched (unmapped) merged headers melt into rows.
    value_pattern: str
    key_column: str = Field(pattern=NAME_RE)
    value_column: str = Field(pattern=NAME_RE)
    value_type: Literal["INTEGER", "REAL", "TEXT"]
    value_parse: Literal["number", "currency", "percent"] | None = None


class TablePlan(CamelModelOpen):
    name: str = Field(pattern=NAME_RE)
    anchor: TableAnchor
    header_rows: int = Field(ge=1, le=3)
    exclude: list[RowRule]
    columns: list[ColumnPlan] = Field(min_length=1)
    unpivot: Unpivot | None = None
    # Canonical column (parse: "date") validated against the slot period.
    period_column: str | None = None
    on_period_mismatch: Literal["keep", "exclude"] = "keep"  # zod .default("keep")


class ParsePlan(CamelModelOpen):
    version: Literal[1]
    tables: list[TablePlan] = Field(min_length=1)


class _ExecReportRowExcluded(CamelModelOpen):
    pattern: str
    count: float
    samples: list[str]


class _ExecReportCoercionFailure(CamelModelOpen):
    column: str
    count: float
    samples: list[str]


class _ExecReportPeriodMismatches(CamelModelOpen):
    count: float
    samples: list[str]


class _ExecReportAnchor(CamelModelOpen):
    sheet: str
    row: float


class _ExecReportTable(CamelModelOpen):
    name: str
    anchor: _ExecReportAnchor | None
    # Byte-exact problem lines; a table with problems produces no rows.
    problems: list[str]
    rows_kept: float
    rows_excluded: list[_ExecReportRowExcluded]
    coercion_failures: list[_ExecReportCoercionFailure]
    period_mismatches: _ExecReportPeriodMismatches | None
    unknown_columns: list[str]


class ExecReport(CamelModelOpen):
    tables: list[_ExecReportTable]


class _PlanPreviewTable(CamelModelOpen):
    name: str
    columns: list[str]
    rows: list[list[str | float | None]]


class PlanPreview(CamelModelOpen):
    tables: list[_PlanPreviewTable]


def validate_parse_plan(plan: dict) -> None:
    """Port of validateParsePlan — structural validity beyond zod.

    Raises ValueError whose message is the first violated rule; messages are
    stable and user-visible. Operates on the plain-dict runtime shape.
    """
    table_names: set[str] = set()
    for t in plan["tables"]:
        name = t["name"]
        if name in table_names:
            raise ValueError(f"duplicate table name: {name}")
        table_names.add(name)
        canon: set[str] = set()
        froms: set[str] = set()
        for c in t["columns"]:
            if c["name"] == "_period":
                raise ValueError(f"reserved column name: {name}._period")
            if c["name"] in canon:
                raise ValueError(f"duplicate column name: {name}.{c['name']}")
            canon.add(c["name"])
            f = re.sub(r"\s+", " ", c["from"].strip().lower())
            if f in froms:
                raise ValueError(f"duplicate column from: {name}.{c['from']}")
            froms.add(f)

        def ref(col: str | None, _canon: set[str] = canon, _name: str = name) -> None:
            if col is not None and col not in _canon:
                raise ValueError(f"unknown column reference: {_name}.{col}")

        for r in t["exclude"]:
            ref(r["column"])
            try:
                plan_pattern(r["pattern"])
            except re.error:
                raise ValueError(f"invalid pattern: {name}.{r['pattern']}") from None

        unpivot = t.get("unpivot")
        if unpivot:
            for k in unpivot["keep"]:
                ref(k)
            try:
                plan_pattern(unpivot["valuePattern"])
            except re.error:
                raise ValueError(f"invalid pattern: {name}.{unpivot['valuePattern']}") from None
            if (
                unpivot["keyColumn"] in canon
                or unpivot["valueColumn"] in canon
                or unpivot["keyColumn"] == unpivot["valueColumn"]
            ):
                raise ValueError(f"unpivot column collides: {name}")

        period_column = t.get("periodColumn")
        if period_column is not None:
            ref(period_column)
            c = next((x for x in t["columns"] if x["name"] == period_column), None)
            if not c or c.get("parse") != "date":
                raise ValueError(f'periodColumn must have parse "date": {name}.{period_column}')


def plan_table_name(family_key: str, plan: dict, logical_name: str) -> str:
    """Warehouse table name for one logical table (mirrors tableNamesFor's single-table rule)."""
    return f"fam_{family_key}" if len(plan["tables"]) == 1 else f"fam_{family_key}__{logical_name}"
